jax/experimental/mesh_utils.py

```diff
@@ -156,7 +192,27 @@ def _jax_devices_order_normalized(
   return jax_devices_array.transpose()
 
 
-def create_device_mesh(mesh_shape: Sequence[int]) -> np.ndarray:
+# jekbradbury's famous trick for creating contiguous submeshes (where available)
+def _transpose_trick(physical_mesh: np.ndarray,
+                     mesh_shape: Sequence[int]) -> np.ndarray:
+  mesh_shape = tuple(mesh_shape)
+  topology = physical_mesh.shape
+  if topology not in _TRANSPOSE_TRICKS:
+    raise ValueError(
+        f"create_device_mesh cannot create contiguous submeshes for "
+        f"physical mesh topology {topology}")
+
+  if mesh_shape not in _TRANSPOSE_TRICKS[topology]:
+    raise ValueError(
+        f"create_device_mesh cannot create contiguous submeshes for "
+        f"mesh_shape {mesh_shape} and physical mesh topology {topology}. "
+        f"Available mesh_shapes: {list(_TRANSPOSE_TRICKS[topology].keys())}")
+
+  return physical_mesh.transpose(*_TRANSPOSE_TRICKS[topology][mesh_shape])
+
+
+def create_device_mesh(mesh_shape: Sequence[int],
+                       contiguous_submeshes: bool = False) -> np.ndarray:
   """Creates a performant device mesh for jax.experimental.maps.mesh.
 
```

Comment:
In the comments for `Returns`, maybe say jax *global* devices? i like how you made it explicit elsewhere that global devices need to be passed in